### ISSUE TITLE
Underwater custom fog colors do not take world data into account

### ISSUE THREAD

**Issue by Builderb0y** [sp614x/optifine#444]
For example, day and night. fogColor is expected to be darker at night, and it is except when underwater. I render water fog manually with a (mostly) hard-coded color so that I can get a more accurate mix of fog colors when looking at things through water. Specifically, I render fog twice. Once for the distance between water and the terrain behind it using fogColor, then a 2nd time for the distance to the water using my own color. (Colors reversed when not in water.) As such, I'm using custom colors to ensure that fogColor doesn't abruptly change when going underwater (simply by using the same [image](http://i.imgur.com/gwaijvJ.png) for fog0.png and underwater.png). This works great during the day, but at night it doesn't modify fogColor to be dark as expected so you just end up with extremely over-brightened colors while underwater ([screenshot](http://i.imgur.com/ULJp8yo.png)).

I'd very much appreciate having an option somewhere to either make underwater fog colors obey the same rules as regular fog, or to ignore water fog colors entirely and just return the regular fog color (which DOES follow the correct rules). The latter of which would probably be more useful in my case, since most custom fog color resource packs that I've seen (including vanilla colors) have different normal vs. underwater fog colors to begin with, so fogColor would change abruptly when entering water regardless of day/night in those packs.

**Comment by sp614x** [comment]
The custom underwater color is not corrected for world time, this is like vanilla. 
In vanilla this is not visible, because the color is very dark (0.02, 0.02, 0.02). 
The color is only corrected for water breathing (lighter). When water breathing is active it is easy to see that the underwater fog is the same during the day and the night.

**Comment by Builderb0y** [comment]
The fact that it's so dark in vanilla is exactly why I'm rendering it manually via shaders to begin with. Dark fog on a bright day looks extremely unnatural in my opinion.

**Comment by sp614x** [comment]
I could correct only the custom underwater color by world time and leave the vanilla color as it is.

**Comment by Builderb0y** [comment]
As far as my shader pack is concerned, I simply need the underwater fog color to match the normal one. This includes day/night, but also the fact that it gets reddened at sunset (and more so when looking in the general direction of the sun). Pretty sure rain strength changes it too.

**Comment by sp614x** [comment]
The normal fog color (air) is due to atmospheric effects (reddening at sunset, day/night, rain, etc).
Reddening the underwater fog color at sunset would be strange.

**Comment by Builderb0y** [comment]
Again, I'm handling underwater fog manually. The fogColor uniform is used for rendering above-water fog. So if you're underwater, and looking up at the surface, and see a mountain above you, it would first apply fogColor fog to the mountain using the distance between the mountain and the surface of the water. Then, it would render the water texture on top of the foggy mountain (when enabled). And lastly, it would apply a hard-coded blue color fog on top of THAT using the distance from you to the surface of the water.

If fogColor becomes dark blue while underwater, then it would apply a dark blue fog to the mountain instead of a bright color, and then a 2nd layer of blue fog on top of that. Additionally, this doesn't just apply to oceans. Standing in a fountain, for example, would make the entire world apply dark blue fog. The goal is to make the world look like [this](http://i.imgur.com/AIjYK6v.png), not like [this](http://i.imgur.com/vDEv00d.png) (blur/refraction effects disabled in those screenshots).

The point is that it doesn't look natural to have that bright of a fog color at night, nor does it look natural to NOT have it reddened at sunset. Fog doesn't magically change color in real life when looking at it through a fish tank, after all.

**Comment by sp614x** [comment]
If I understand correctly you want to deactivate the underwater fog color and always use the normal fog color? In this case the uniform fogColor will not change when going underwater.

**Comment by Builderb0y** [comment]
Correct. I like the idea of having a 2nd uniform too, since that would allow for biome-specific underwater fog colors. And I already have a config option for old optifine versions as a fallback incase that uniform does not exist. Keep in mind though that the underwater fog color will also have to not change when going above water, since in that case underwater fog color will be applied first, then the water texture, then the normal fog color. In other words, both fog colors are always used regardless of weather or not you're actually in water.
